packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.tsx

```diff
@@ -86,6 +85,43 @@ export const SelectColumnsButton = props => {
         }
     };
 
+    const handleMove = (index1, index2) => {
+        const index1Pos = availableColumns.findIndex(
+            // eslint-disable-next-line eqeqeq
+            field => field.index == index1
+        );
+        const index2Pos = availableColumns.findIndex(
+            // eslint-disable-next-line eqeqeq
+            field => field.index == index2
+        );
+        if (index1Pos === -1 || index2Pos === -1) {
+            return;
+        }
+        let newAvailableColumns;
+        if (index1Pos > index2Pos) {
+            newAvailableColumns = [
+                ...availableColumns.slice(0, index2Pos),
+                availableColumns[index1Pos],
+                ...availableColumns.slice(index2Pos, index1Pos),
+                ...availableColumns.slice(index1Pos + 1),
+            ];
+        } else {
+            newAvailableColumns = [
+                ...availableColumns.slice(0, index1Pos),
+                ...availableColumns.slice(index1Pos + 1, index2Pos + 1),
+                availableColumns[index1Pos],
+                ...availableColumns.slice(index2Pos + 1),
+            ];
+        }
+        setAvailableColumns(newAvailableColumns);
+        setColumns(columns =>
+            newAvailableColumns
+                .filter(column => !omit?.includes(column.source))
```

Comment:
I don't think you should use `omit` here. Reordering the columns shouldn't re-apply the omit param IMO.